pulser-core/pulser/backend/state.py

```diff
@@ -0,0 +1,236 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Defines the abstract base class for a quantum state."""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections import Counter
+from collections.abc import Mapping, Sequence
+from itertools import product
+from typing import Any, Generic, Literal, Protocol, Type, TypeVar, Union
+
+import numpy as np
+
+from pulser.channels.base_channel import States
+
+Eigenstate = Union[States, Literal["0", "1"]]
+
+ArgScalarType = TypeVar("ArgScalarType")
+ReturnScalarType = TypeVar("ReturnScalarType")
+StateType = TypeVar("StateType", bound="State")
+
+
+class _ProbabilityType(Protocol):
+    """A protocol for probability types.
+
+    Defines only the methods needed to correctly type hint State.
+    """
+
+    def __float__(self) -> float: ...
+
+    def __add__(self, other: Any) -> float: ...
+
+
+class State(ABC, Generic[ArgScalarType, ReturnScalarType]):
+    """Base class enforcing an API for quantum states.
+
+    Each backend will implement its own type of state and the methods below.
+    """
+
+    _eigenstates: Sequence[Eigenstate]
+
+    @property
+    @abstractmethod
+    def n_qudits(self) -> int:
+        """The number of qudits in the state."""
+        pass
+
+    @property
+    def eigenstates(self) -> tuple[Eigenstate, ...]:
+        """The eigenstates that form a qudit's eigenbasis.
+
+        The order of the states should match the order in a numerical (ie
+        state vector or density matrix) representation.
+        For example, with eigenstates ("a", "b", ...),  "a" will be associated
+        to eigenvector (1, 0, ...), "b" to (0, 1, ...) and so on.
+        """
+        return tuple(self._eigenstates)
+
+    @property
+    def basis_states(self) -> tuple[str, ...]:
+        """The basis states combinations, in order."""
+        return tuple(
+            map(
+                "".join,
+                product("".join(self.eigenstates), repeat=self.n_qudits),
+            )
+        )
+
+    @property
+    def qudit_dim(self) -> int:
+        """The dimensions (ie number of eigenstates) of a qudit."""
+        return len(self.eigenstates)
+
+    @abstractmethod
+    def overlap(self: StateType, other: StateType, /) -> ReturnScalarType:
+        """Compute the overlap between this state and another of the same type.
+
+        Generally computes Tr[AB] for mixed states A and B, which
+        corresponds to |<a|b>|^2 for pure states A=|a><a| and B=|b><b|.
+
+        Args:
+            other: The other state.
+
+        Returns:
+            The overlap between the two states.
+        """
+        pass
+
+    @abstractmethod
+    def probabilities(
```

Comment:
This also does not scale well with qubit number. Say you create a 100 qubit hadamard state, which is quite trivial to do for an MPS at least.  This has the uniform distribution on the bitstrings, and these probabilities cannot be represented in any floating point currently in use since the number `2^{-100}` is well below machine precision. Conversely sample methods bases on a Markov chain (such as we currently use in emu-mps) would only deal with numbers of order unity and sample from the underlying distribution correctly.

Reply:
Ok, that's good to know. So bascially what you're saying is that we need to have the whole `sample()` method be the `@abstractmethod` and we can't have any shared logic in `State`, right?

Reply:
Moved out of `State`